yjit/src/invariants.rs

```diff
@@ -397,87 +397,58 @@ pub extern "C" fn rb_yjit_tracing_invalidate_all()
         return;
     }
 
-    todo!("rb_yjit_tracing_invalidate_all");
-/*
+    use crate::asm::x86_64::jmp_ptr;
+
     // Stop other ractors since we are going to patch machine code.
-    RB_VM_LOCK_ENTER();
-    rb_vm_barrier();
-
-    // Make it so all live block versions are no longer valid branch targets
-    rb_objspace_each_objects(tracing_invalidate_all_i, NULL);
-
-    // Apply patches
-    const uint32_t old_pos = cb->write_pos;
-    rb_darray_for(global_inval_patches, patch_idx) {
-        struct codepage_patch patch = rb_darray_get(global_inval_patches, patch_idx);
-        cb.set_pos(patch.inline_patch_pos);
-        uint8_t *jump_target = cb_get_ptr(ocb, patch.outlined_target_pos);
-        jmp_ptr(cb, jump_target);
-    }
-    cb.set_pos(old_pos);
-
-    // Freeze invalidated part of the codepage. We only want to wait for
-    // running instances of the code to exit from now on, so we shouldn't
-    // change the code. There could be other ractors sleeping in
-    // branch_stub_hit(), for example. We could harden this by changing memory
-    // protection on the frozen range.
-    RUBY_ASSERT_ALWAYS(yjit_codepage_frozen_bytes <= old_pos && "frozen bytes should increase monotonically");
-    yjit_codepage_frozen_bytes = old_pos;
-
-    cb_mark_all_executable(ocb);
-    cb_mark_all_executable(cb);
-    RB_VM_LOCK_LEAVE();
-*/
-}
+    with_vm_lock(src_loc!(), || {
+        // Make it so all live block versions are no longer valid branch targets
+        unsafe { rb_yjit_for_each_iseq(Some(invalidate_all_blocks_for_tracing)) };
+
+        extern "C" fn invalidate_all_blocks_for_tracing(iseq: IseqPtr) {
+            if let Some(payload) = unsafe { load_iseq_payload(iseq) } {
+                // C comment:
+                //   Leaking the blocks for now since we might have situations where
+                //   a different ractor is waiting for the VM lock in branch_stub_hit().
+                //   If we free the block that ractor can wake up with a dangling block.
+                //
+                // Deviation: since we ref count the the blocks now, we might be deallocating and
+                //   not leak the block.
+                //
+                // Empty all blocks on the iseq so we don't compile new blocks that jump to the
+                // invalidated region.
+                let version_map = payload.take_version_map();
+                for versions in &version_map {
+                    for blockref in versions {
+                        block_assumptions_free(&blockref);
+                    }
+                }
```

Comment:
Should maybe put this code in `src/core.rs` to avoid making the `VersionMap` public? Or maybe create a method `IseqPayload::clear_all_blocks()` ?